wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt

```diff
@@ -211,6 +220,154 @@ internal sealed class GrpcMethod<S : Any, R : Any>(
+          get() = call!!.timeout()
+
+        override fun cancel() {
+          call?.cancel()
+        }
+
+        override suspend fun execute(request: S): R {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          return suspendCancellableCoroutine { continuation ->
+            continuation.invokeOnCancellation {
+              cancel()
+            }
+            call!!.enqueue(object : Callback {
+              override fun onFailure(call: Call, e: IOException) {
+                continuation.resumeWithException(e)
+              }
+
+              override fun onResponse(call: Call, response: Response) {
+                try {
+                  response.use {
+                    response.messageSource(responseAdapter).use { reader ->
+                      while (true) {
+                        val message = reader.read() ?: break
+                        continuation.resume(message)
+                      }
+
+                      val exception = response.grpcStatusToException()
+                      if (exception != null) continuation.resumeWithException(exception)
+                    }
+                  }
+                } catch (e: IOException) {
+                  continuation.resumeWithException(e)
+                }
+              }
+            })
+          }
+        }
+
+        override fun executeBlocking(request: S): R {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          val messageSource = BlockingMessageSource(call!!, responseAdapter)
+          call!!.enqueue(messageSource.readFromResponseBodyCallback())
+
+          messageSource.use {
+            return messageSource.read() ?: throw ProtocolException("required message not sent")
+          }
+        }
+
+        override fun enqueue(request: S, callback: com.squareup.wire.GrpcCall.Callback<S, R>) {
+          val requestBody = newRequestBody(requestAdapter, request)
+          call = grpcClient.newCall(path, requestBody)
+
+          val messageSource = BlockingMessageSource(call!!, responseAdapter)
+          call!!.enqueue(messageSource.readFromResponseBodyCallback())
+
+          messageSource.use {
+            try {
+              val response =
+                  messageSource.read() ?: throw ProtocolException("required message not sent")
+              callback.onSuccess(this, response)
+            } catch (e: IOException) {
+              callback.onFailure(this, e)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  class GrpcStreamingCall<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(
+      grpcClient: GrpcClient,
+      args: Array<Any>
+    ): com.squareup.wire.GrpcStreamingCall<S, R> {
```

Comment:
Tempting to put these inside-out, with the `GrpcStreamingCall` subtype being the named outer class